.github/workflows/ci.yaml

```diff
@@ -86,60 +51,156 @@ jobs:
+        run: docker push $REGISTRY/$IMAGE_NAME:${GITHUB_SHA::7} && docker push $REGISTRY/$IMAGE_NAME:latest
+
+      - name: push Docker image on PR event
+        if: github.event_name == 'pull_request'
+        run: |
+          PR_SHA7=${{github.event.pull_request.head.sha}}
+          docker push $REGISTRY/$IMAGE_NAME:${PR_SHA7::7} && docker push $REGISTRY/$IMAGE_NAME:latest
+
+      #---Build Docker image, Login & Publish to Docker Packages - WITH 3RD PARTY ACTIONS---
+      #- name: Log in to the Container registry
+      #  uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+      #  with:
+      #    registry: $REGISTRY
+      #    username: ${{ github.actor }} #${{ github.repository_owner }}
+      #    password: ${{ secrets.GITHUB_TOKEN }}
       
+      #- name: Extract metadata (tags, labels) for Docker
+      #  id: meta
+      #  uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
+      #  with:
+      #    images: $REGISTRY/$IMAGE_NAME
+            
+      #- name: Build and push Docker image
+      #  uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
+      #  with:
+      #    context: ./angular
+      #    push: true
+      #    tags: ${{ steps.meta.outputs.tags }}
+      #    labels: ${{ steps.meta.outputs.labels }}
+  
   stats_job:
-    runs-on: ubuntu-latest
-    needs: main_job
+    timeout-minutes: 30
+    runs-on: ubuntu-20.04
+    if: ${{ always() }}
+    needs: [build_job, test_job]
     steps:
-      - name:  Extract job starting & ending time
+      - name:  Extract TEST and BUILD job starting & ending time
         id: execution_timestamps
         run: |
-          echo "::set-output name=starttime::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].started_at')" && echo "::set-output name=endtime::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].completed_at')"
-      
+          echo "::set-output name=starttime_test::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].started_at')" && echo "::set-output name=endtime_test::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].completed_at')"
+          echo "::set-output name=starttime_build::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[1].started_at')" && echo "::set-output name=endtime_build::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[1].completed_at')"
```

Comment:
1. that code is very hard to read. Pls improve readability
2. what if you decide to break the pipeline into more jobs? Your finish time calc would be incorrect.

Reply:
changed time calculation approach